File: src/factories/playwrightFactory.ts
import { chromium, Browser, BrowserContext, Page } from '@playwright/test';
import { PlaywrightConfig } from '../config/playwrightConfig';

export interface PlaywrightSession {
  browser: Browser;
  context: BrowserContext;
  page: Page;
}

/**
 * Creates local and BrowserStack Playwright sessions.
 */
export class PlaywrightFactory {
  public static async getBrowser(config: PlaywrightConfig, testName?: string): Promise<PlaywrightSession> {
    config.validateOrThrow();

    const browser = config.isBrowserStackExecution()
      ? await this.connectBrowserStack(config, testName)
      : await this.launchLocal(config);

    const context = await browser.newContext({
      ignoreHTTPSErrors: true,
      viewport: { width: 1920, height: 1080 }
    });

    const page = await context.newPage();
    page.setDefaultTimeout(config.defaultTimeoutMs);

    return {
      browser,
      context,
      page
    };
  }

  private static async launchLocal(config: PlaywrightConfig): Promise<Browser> {
    const channel = config.browserType === 'edge' ? 'msedge' : 'chrome';

    return chromium.launch({
      headless: config.headless,
      channel
    });
  }

  private static async connectBrowserStack(config: PlaywrightConfig, testName?: string): Promise<Browser> {
    const browser = config.browserType === 'bs_edge' ? 'edge' : 'chrome';
    const caps = {
      browser,
      browser_version: 'latest',
      os: 'Windows',
      os_version: '11',
      name: testName ?? 'Playwright TypeScript Session',
      build: `AutoBots Playwright TS Migration - ${new Date().toISOString().slice(0, 10)}`,
      'browserstack.username': config.browserStackUserName,
      'browserstack.accessKey': config.browserStackAccessKey
    };

    const endpoint = `${config.connectionString}?caps=${encodeURIComponent(JSON.stringify(caps))}`;
    return chromium.connect(endpoint);
  }
}

File: src/AutobotsPlaywrightFramework/Factories/PlaywrightFactory.cs
using System.Text;
using System.Text.Json;
using AutobotsPlaywrightFramework.Configurations;
using Microsoft.Playwright;
using NUnit.Framework;

namespace AutobotsPlaywrightFramework.Factories;

/// <summary>
/// Factory used to initialize local and BrowserStack Playwright sessions.
/// </summary>
public static class PlaywrightFactory
{
    /// <summary>
    /// Creates a fully initialized Playwright session using runsettings-backed configuration.
    /// </summary>
    /// <param name="browserName">The browser option from run settings.</param>
    public static Task<PlaywrightSession> GetBrowserAsync(string browserName)
    {
        var config = PlaywrightConfig.FromTestContext();
        config.ValidateOrThrow();
        return GetBrowserAsync(browserName, config, TestContext.CurrentContext.Test.Name);
    }

    /// <summary>
    /// Creates a fully initialized Playwright session based on run settings.
    /// </summary>
    /// <param name="browserName">The browser option from run settings.</param>
    /// <param name="config">Runtime Playwright configuration.</param>
    /// <param name="testName">Optional test name for BrowserStack session naming.</param>
    public static async Task<PlaywrightSession> GetBrowserAsync(string browserName, PlaywrightConfig config, string? testName = null)
    {
        ArgumentNullException.ThrowIfNull(config);

        var playwright = await Microsoft.Playwright.Playwright.CreateAsync().ConfigureAwait(false);
        IBrowser browser = await CreateBrowserAsync(playwright, browserName, config, testName).ConfigureAwait(false);

        var context = await browser.NewContextAsync(new BrowserNewContextOptions
        {
            IgnoreHTTPSErrors = true,
            ViewportSize = new ViewportSize { Width = 1920, Height = 1080 }
        }).ConfigureAwait(false);

        var page = await context.NewPageAsync().ConfigureAwait(false);
        page.SetDefaultTimeout(config.DefaultTimeoutMs);

        return new PlaywrightSession(playwright, browser, context, page);
    }

    private static Task<IBrowser> CreateBrowserAsync(IPlaywright playwright, string browserName, PlaywrightConfig config, string? testName)
    {
        return browserName.ToLowerInvariant() switch
        {
            "chrome" => playwright.Chromium.LaunchAsync(new BrowserTypeLaunchOptions
            {
                Channel = "chrome",
                Headless = false
            }),
            "edge" => playwright.Chromium.LaunchAsync(new BrowserTypeLaunchOptions
            {
                Channel = "msedge",
                Headless = false
            }),
            "bs_chrome" => ConnectBrowserStackAsync(playwright, "chrome", config, testName),
            "bs_edge" => ConnectBrowserStackAsync(playwright, "edge", config, testName),
            _ => throw new ArgumentOutOfRangeException(nameof(browserName), browserName, "Unsupported browser option.")
        };
    }

    private static Task<IBrowser> ConnectBrowserStackAsync(IPlaywright playwright, string browserName, PlaywrightConfig config, string? testName)
    {
        var capabilities = new Dictionary<string, object?>
        {
            ["browser"] = browserName,
            ["browser_version"] = "latest",
            ["os"] = "Windows",
            ["os_version"] = "11",
            ["name"] = testName ?? TestContext.CurrentContext.Test.Name,
            ["build"] = $"AutoBots Playwright Migration - {DateTime.UtcNow:yyyyMMdd}",
            ["browserstack.username"] = config.BrowserStackUserName,
            ["browserstack.accessKey"] = config.BrowserStackAccessKey
        };

        var capabilitiesJson = JsonSerializer.Serialize(capabilities);
        var encodedCaps = Convert.ToBase64String(Encoding.UTF8.GetBytes(capabilitiesJson));
        var endpoint = $"{config.ConnectionString}?caps={encodedCaps}";

        return playwright.Chromium.ConnectAsync(endpoint);
    }
}

/// <summary>
/// Encapsulates Playwright runtime objects for a test session.
/// </summary>
public sealed record PlaywrightSession(
    IPlaywright Playwright,
    IBrowser Browser,
    IBrowserContext Context,
    IPage Page);
